python/mage/tgn/definitions/tgn.py

```diff
@@ -0,0 +1,1213 @@
+                            timestamp=event.timestamp,
+                            dest_memory=memory.get_node_memory(event.source),
+                            source=event.dest,
+                            edge_features=edge_features[event.edge_indx],
+                            delta_time=torch.tensor(
+                                np.array(event.timestamp).astype("float"),
+                                requires_grad=True,
+                            )
+                            - memory.get_last_node_update(event.dest),
+                        )
+                    )
+                else:
+                    raise Exception(f"Event Type not supported {type(event)}")
+        return raw_messages
+
+    def _aggregate_messages(
+        self,
+        processed_messages: Dict[int, List[torch.Tensor]],
+        aggregator_function: MessageAggregator,
+    ) -> Dict[int, torch.Tensor]:
+        # todo change so that it returns for aggregated messages
+        #                               [[]
+        #                                []], shape=(number_of_nodes, aggregated_length)
+        aggregated_messages = {node: None for node in processed_messages}
+        for node in processed_messages:
+            aggregated_messages[node] = aggregator_function(processed_messages[node])
+        return aggregated_messages
+
+    def _update_memory(self, messages, memory, memory_updater) -> None:
+        # todo change to do all updates at once
+        for node in messages:
+            updated_memory = memory_updater(
+                (messages[node], memory.get_node_memory(node))
+            )
+
+            # use flatten to get (memory_dim,)
+            updated_memory = torch.flatten(updated_memory)
+            memory.set_node_memory(node, updated_memory)
+
+    def _form_computation_graph(
+        self, nodes: np.array, timestamps: np.array
+    ) -> (
+        List[List[Tuple[int, int]]],
+        List[Dict[Tuple[int, int], int]],
+        List[List[int]],
+        List[List[int]],
+    ):
+        node_layers = [[(n, t) for (n, t) in zip(nodes, timestamps)]]
+        edge_layers = []
+        timestamp_layers = []
+
+        for _ in range(self.num_of_layers):
+            prev = node_layers[-1]
+            cur_arr = [(n, v) for (n, v) in prev]
+
+            node_arr = []
+            for (v, t) in cur_arr:
+                (
+                    neighbors,
+                    edge_idxs,
+                    timestamps,
+                ) = self.temporal_neighborhood.get_neighborhood(
+                    v, t, self.num_neighbors
+                )
+                node_arr.extend([(ni, ti) for (ni, ti) in zip(neighbors, timestamps)])
+            cur_arr.extend(node_arr)
+            node_arr = list(set(cur_arr))
+            node_layers.append(node_arr)
+
+        node_layers.reverse()
+
+        # this mapping will be later used to reference node features and edge features and time features
```

Comment:
Update on sampling

Reply:
I am not sure I understand this comment?